emu/src/arm7tdmi.rs

```diff
@@ -152,38 +152,32 @@ impl Arm7tdmi {
         let reg_list = op_code.get_bits(0..=15);
 
         let memory_base = self.registers.register_at(rn.try_into().unwrap());
-        let alignment = 4; // Since are word, the alignment is 4.
         let mut address = memory_base;
-        let change_address = |address: u32| {
-            if up_down {
-                address.wrapping_add(alignment)
-            } else {
-                address.wrapping_sub(alignment)
-            }
-        };
 
         if load_store {
-            for reg_destination in 0..16 {
-                if reg_list.is_bit_on(reg_destination) {
-                    if pre_post {
-                        address = change_address(address);
-                    }
-
-                    let part_0: u32 = self.memory.read_at(address).try_into().unwrap();
-                    let part_1: u32 = self.memory.read_at(address + 1).try_into().unwrap();
-                    let part_2: u32 = self.memory.read_at(address + 2).try_into().unwrap();
-                    let part_3: u32 = self.memory.read_at(address + 3).try_into().unwrap();
-                    let value: u32 = part_3 << 24_u32 | part_2 << 16_u32 | part_1 << 8_u32 | part_0;
-                    self.registers
-                        .set_register_at(reg_destination.try_into().unwrap(), value);
-
-                    if !pre_post {
-                        address = change_address(address);
-                    }
-                }
-            }
+            let transfer = |arm: &mut Self, address: u32, reg_destination: usize| {
+                let part_0: u32 = arm.memory.read_at(address).try_into().unwrap();
+                let part_1: u32 = arm.memory.read_at(address + 1).try_into().unwrap();
+                let part_2: u32 = arm.memory.read_at(address + 2).try_into().unwrap();
+                let part_3: u32 = arm.memory.read_at(address + 3).try_into().unwrap();
+                let v = part_3 << 24_u32 | part_2 << 16_u32 | part_1 << 8_u32 | part_0;
+                arm.registers.set_register_at(reg_destination, v);
+            };
+
+            self.exec_data_trasfer(reg_list, pre_post, &mut address, up_down, transfer);
         } else {
-            todo!()
+            let transfer = |arm: &mut Self, address: u32, reg_source: usize| {
+                let value = arm.registers.register_at(reg_source);
+                arm.memory.write_at(address, value.get_bits(0..=7) as u8);
+                arm.memory
+                    .write_at(address + 1, value.get_bits(8..=15) as u8);
+                arm.memory
+                    .write_at(address + 2, value.get_bits(16..=23) as u8);
+                arm.memory
+                    .write_at(address + 3, value.get_bits(24..=31) as u8);
+            };
+
+            self.exec_data_trasfer(reg_list, pre_post, &mut address, up_down, transfer);
```

Comment:
What do you think about moving this logic to `InternalMemory`? Maybe changing the trait `IoDevice`. We could introduce functions like `write_32_bits` `read_32_bits` or change the current `write_at` and `read_at` to support the amount of bytes to read/write.

Reply:
I already thought on it, I think I will do in separated PR, I'm only thinking cleanest solution to do so :)

Reply:
Fair enough, can we create an issue so we don't forget? :D

Reply:
I can do it :)